src/components/tables/Table.tsx

```diff
@@ -0,0 +1,337 @@
+import { ITableDispatchProps } from './TableConnected';
+import { getTableChildComponentId } from './TableUtils';
+import { TableChildComponent } from './TableConstants';
+import { JSXRenderable } from '../../utils/JSXUtils';
+import { convertUndefinedAndNullToEmptyString } from '../../utils/FalsyValuesUtils';
+import { TableCollapsibleRowConnected } from './TableCollapsibleRowConnected';
+import { LoadingConnected } from '../loading/LoadingConnected';
+import { LoadingTableConnected } from '../loading/LoadingTableConnected';
+import { INavigationChildrenProps } from '../navigation/Navigation';
+import { NavigationConnected } from '../navigation/NavigationConnected';
+import { IDropdownOption } from '../dropdownSearch/DropdownSearch';
+
+export interface IData {
+  id: string;
+  [attribute: string]: any;
+};
+
+export interface ITableRowData {
+  [id: string]: {
+    id: string;
+    [attribute: string]: any;
+  };
+};
+
+export type IAttributeFormatter = (attributeValue: any, attributeName?: string) => JSXRenderable;
+export type IAttributeNameFormatter = (attributeName: string) => string;
+
+export interface ITableHeadingAttribute {
+  attributeName: string;
+  titleFormatter: (attributeName: string) => string;
+  sort?: boolean;
+  sortByMethod?: (attributeValue: any) => string;
+  attributeFormatter?: IAttributeFormatter;
+}
+
+export interface ITablePredicate {
+  props: IDropdownSearchProps;
+  attributeName: string;
+  attributeNameFormatter: IAttributeNameFormatter;
+};
+
+export interface ITableOwnProps extends React.ClassAttributes<Table> {
+  id: string;
+  initialTableData: ITableData;
+  headingAttributes: ITableHeadingAttribute[];
+  collapsibleFormatter?: (tableRowData: ITableRowData, props: ITableProps) => JSXRenderable;
+  actionBar?: IActionBarProps;
+  getActions?: (rowData?: ITableRowData, props?: ITableProps) => IActionOptions[];
+  blankSlates: {
+    noResults: IBlankSlateProps;
+    noResultsOnFilterOrPredicates?: IBlankSlateProps;
+    noResultsOnError?: IBlankSlateProps;
+  };
+  filter?: IFilterBoxProps;
+  filterMethod?: (attributeValue: any, props: ITableOwnProps) => boolean;
+  predicates?: ITablePredicate[];
+  navigation?: INavigationChildrenProps;
+  noNavigation?: boolean;
+  lastUpdatedLabel?: string;
+
+  // modes
+  serverMode?: {
+    url: (tableState?: ITableState, ownProps?: ITableOwnProps) => string;
+    rawDataToTableData: (data: any, ownProps?: ITableOwnProps, tableState?: ITableState) => ITableData;
+  };
+  customMode?: {
+    thunkActionCreator: (tableOwnProps?: ITableOwnProps, shouldResetPage?: boolean) => ((dispatch: any, getState?: () => any) => void);
+  };
+};
+
+export interface ITableStateProps {
+  tableState?: ITableState;
+};
+
+export interface ITableProps extends ITableOwnProps, ITableStateProps, Partial<ITableDispatchProps> { }
+
+export class Table extends React.Component<ITableProps, any> {
+  componentDidMount() {
+    if (this.props.onDidMount) {
```

Comment:
Everywhere else we use onRender and onDestroy

Reply:
i would tend to say it's good to define a dispatch prop that reflects where you use it afterward, here i initialize the table state in componentDidMount because i want its children to be in the state before it being in the state (and they do mount before the parent). https://github.com/facebook/react/blob/v15.0.1/docs/docs/ref-03-component-specs.md#mounting-componentdidmount